packages/imperative/src/rest/src/client/AbstractRestClient.ts

```diff
@@ -220,15 +221,29 @@ export abstract class AbstractRestClient {
      */
     protected lastByteReceived: number = 0;
 
+    /**
+     * Specifies the order in which credentials will be used when multiple
+     * types of credentials exist in the zowe client configuration.
+     * @private
+     * @type {ICredOrder}
+     * @memberof AbstractRestClient
+     */
+    private mCredOrder: ICredOrder;
+
     /**
      * Creates an instance of AbstractRestClient.
      * @param {AbstractSession} mSession - representing connection to this api
+     * @param {ICredOrder} credOrder
+     *          Specifies the order in which credentials will be used when multiple
+     *          types of credentials exist in the zowe client configuration.
+     *          When unspecified, a token will be used before user & password.
      * @memberof AbstractRestClient
      */
-    constructor(private mSession: AbstractSession) {
+    constructor(private mSession: AbstractSession, credOrder = ICredOrder.TOKEN_OVER_PASSWORD) {
```

Comment:
without much thought, I "assumed" this option would have been implemented here: https://github.com/zowe/zowe-cli/blob/master/packages/imperative/src/rest/src/session/doc/ISession.ts

do you think this is simply an alternative place to add this new option?

Reply:
AbstractRestClient is only called from one file (RestClient). Thus, the correction can be made in 2 existing files (and 1 new file). No other zowe code should see any potential change or capabilities. It would just correct something that was broken.
 
ISession is used in over a dozen other Zowe files. I am a little concerned about how widely visible this change would become.

We have had requests to officially change the credential order of precedence. To do that properly, we may have to introduce the ability to control the order of password, token, certificate, and Personal-Access-Token, (and maybe something else?) Further, we may have to define a way for end-users to set the precedence differently for different sessions. We have not thought through these requirements well enough to start down any particular implementation path.

I suppose that by making a new ISession property optional, it would not break other functions. Further, if RestClient **ALWAYS** sets that property, and AbstractRestClient is the only class to read that property, it could just as effectively revert the breaking change and limit the change to only the broken class. 

Your comment has made me think about making an ISession property more complex so that it might accommodate more than just password and token. A more complex property might provide the means to control the order of more credential types in the future, even though we would only implement password and token now. 

I will try out some of my thoughts and update this pull request if they pan out.

Reply:
Update: The only way to make this alternative non-breaking is for the AbstractRestClient constructor to **Always** set the Session property to prefer "token". Otherwise a direct caller of AbstractRestClient would have to set the new ISession property before calling AbstractRestClient.constructor. The RestClient already overrides this AbstractRestClient behavior to prefer password, so this should be ok.

Reply:
Hi @dkelosky - I made revisions as described above. Please re-review.

Reply:
Hey @gejohnston, thanks for considering this approach. It feels a little more natural to me. Hopefully others feel the same 😃